### PR TITLE
lesspipe: validate and add script dependencies

### LINKED ISSUE #216087: less / lesspipe: no strings command when decompressing brotli file.
### Describe the bug

```
$ less log.br
==> (lesspipe 2.07) append : to filename to view the original brotli file
/nix/store/ffq874rlpwjgj2ix6varyj483bbn05mb-lesspipe-2.07/bin/.lesspipe.sh-wrapped: line 195: strings: command not found
```
### Notify maintainers
@martijnvermaat 

```
 nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.10-bcachefs-5fc458d, NixOS, 23.05 (Stoat), 23.05pre452935.724bfc08923`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.2`
 - channels(daniel): `"home-manager"`
 - channels(root): `"nixos, nixos-unstable, nixpkgs-master"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
 ```

### PR DESCRIPTION
**Pull request by anund**
## Description of changes

Default `less` generates an warning looking for `strings` in nixos when opening certain files. This happens because less reads `LESSOPEN` environment variable and invokes `lesspipe` which expects [many things](https://github.com/wofr06/lesspipe#3-required-programs) to be available on the `PATH`. The less module defines a default value for the [LESSOPEN](https://github.com/NixOS/nixpkgs/blob/964006733a3380811d2e6c5870a8dc42f681429f/nixos/modules/programs/less.nix#L95) environment variable. The [environment.nix](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/environment.nix#L29) file enables the less module by default. `lesspipe` references `strings` [here](https://github.com/wofr06/lesspipe/blob/lesspipe/lesspipe.sh#L597).

Use resholve to wrap `lesspipe.sh` and `lesscomplete` mostly avoiding accident external references and replace reference to `strings` manually as it's inside an array. 

fixes #216087

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by StillerHarpo** [comment]
This works for me. I also had non empty text files which were shown as a lot of empty lines. With this patch they are also shown correct now.

I think @martijnvermaat ins't active on github anymore (They have no activity since July 2021). So we need a new maintainer.

**Review by StillerHarpo — APPROVED** [review]

File: pkgs/tools/misc/lesspipe/default.nix [hunk 1]
```diff
@@ -28,9 +46,61 @@ stdenv.mkDerivation rec {
   installFlags = [ "DESTDIR=$(out)" ];
 
   postInstall = ''
-    for f in lesspipe.sh lesscomplete; do
-      wrapProgram "$out/bin/$f" --prefix-each PATH : "${lib.makeBinPath [ file gnused procps ]}"
-    done
+    # resholve doesn't see strings in an array definition
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/lesspipe/default.nix:49]: @abathur fyi

File: pkgs/tools/misc/lesspipe/default.nix [hunk 2]
```diff
@@ -28,9 +46,61 @@ stdenv.mkDerivation rec {
...
+    substituteInPlace $out/bin/lesspipe.sh --replace 'nodash strings' "nodash ${binutils}/bin/strings"
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/lesspipe/default.nix:50]: I think we should use the unwrapped binutils to have less dependencies.

File: pkgs/tools/misc/lesspipe/default.nix (hunk 1 above)
Comment (anund) [reply pkgs/tools/misc/lesspipe/default.nix:49]: https://github.com/wofr06/lesspipe/blob/63872f3214606e6e7d11616097cf6e3ffa54fc78/lesspipe.sh#L607 the specific case in context

Comment (abathur) [reply pkgs/tools/misc/lesspipe/default.nix:49]: Speaking very broadly, this is out in the ~10% of dynamic behavior that is likely to remain outside of resholve's purview. 

We can either either tell resholve the value of the variable at execution time, or it has to have to be capable of deriving it from the script. 

In order to derive it from the script itself, it basically needs to a full shell execution model in order to figure out what the values should be. This can seem trivial in very narrow cases, but a general solution is actually pretty tricky--the package-time and run-time environments can differ in complex  ways. resholve would basically need to evaluate the script comprehensively enough to both know what the value should be, _and_ be 100% confident that the runtime environment couldn't affect it.

In other words, it's only possible to peel off very thin slices of genuinely dynamic behavior within resholve. It might be able to address functions that execute $N or $@ without any shifting or other variable-mutating shenanigans), but it gets tricky quickly beyond that point.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Can you take a look at the darwin failure?

```
[resholve context] RESHOLVE_INPUTS=/nix/store/dsvdw81x1nzb0c2lr5shywmwffl8dw1s-coreutils-9.4/bin:/nix/store/g2qniscln5apisz3x1kc9cm3pfpsvk58-file-5.45/bin:/nix/store/i55qnx9sxzbgh9r10w8pwj7p5sxbxhzn-gnugrep-3.11/bin:/nix/store/hjkvi2f3vnvm2b7yvnk5vya0fwn0cfb1-gnused-4.9/bin:/nix/store/v3ghbbzjg8vxl3asgj5j5dxjmpbafbw8-gnutar-1.35/bin:/nix/store/jsd6ghbb4x1clbiz2nzbxlxf1whjnx5v-libiconv-50/bin:/nix/store/0kfwzwrp3hw8hdn9zijvrpc91rqsh5ic-procps-1003.1-2008/bin:/nix/store/ng2flfad5y4smcadfkmsiws9izb1f7gj-ncurses-6.4/bin
[resholve context] RESHOLVE_INTERPRETER=/nix/store/d2khhk4xpwpk9khvhp8n6r3k9si03ax1-bash-5.2-p21/bin/bash
[resholve context] RESHOLVE_KEEP='$prog $c1 $c2 $c3 $c4 $c5 $cmd $colorizer $HOME'
[resholve context] /nix/store/yld6gqlkilq709i9njdzl054f4dvh09a-resholve-0.9.1/bin/resholve --overwrite bin/lesspipe.sh
  locale=$(locale|grep LC_CTYPE|tr -d '"') || locale=utf-8
           ^~~~~~
'/nix/store/jyzdmbrkm2k5n8a4njcf8q2s8j3vz1fi-lesspipe-2.11/bin/lesspipe.sh':799: Couldn't resolve command 'locale'
/nix/store/ch57cwfzm3q792wsllq4f05zr9nrfc69-stdenv-darwin/setup: line 140: pop_var_context: head of shell_variables not a function context
error: builder for '/nix/store/1kyisswj1ghi7xk03iybsvv7gkjin547-lesspipe-2.11.drv' failed with exit code 3;
```

File: pkgs/tools/misc/lesspipe/default.nix [hunk 3]
```diff
@@ -28,9 +47,62 @@ stdenv.mkDerivation rec {
...
+        coreutils
+        file
+        gnugrep
+        gnused
+        gnutar
+        iconv
+        procps
+        ncurses
+        locale
+      ];
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/lesspipe/default.nix:66]:
```suggestion
      ] ++ lib.optional stdenv.isDarwin [
        locale
      ];
```

I tested the new commit on my system and got the following diff

```
locale-glibc: ∅ → 2.38-27
```

which means locale is otherwise not used on my system. I think this is something darwin specific.

If there is still something missing after that, we should ask in the darwin channel for help.

**Comment by SuperSandro2000** [comment]
@ofborg build lesspipe

**Comment by SuperSandro2000** [comment]
Can you rebase to fix the merge conflict? Thanks!

File: pkgs/tools/misc/lesspipe/default.nix (hunk 3 above)
Comment (anund) [reply pkgs/tools/misc/lesspipe/default.nix:66]: https://github.com/wofr06/lesspipe/blob/lesspipe/lesspipe.sh#L799 is the usage site. Do you have lesspipe disabled on your system? I can't see how it would would without replacement.

Comment (anund) [reply pkgs/tools/misc/lesspipe/default.nix:66]: I can see `locale` being dragged in for other reasons on non-darwin configurations.

Comment (anund) [reply pkgs/tools/misc/lesspipe/default.nix:66]: Curious:
```bash
$ bash

$ a=$(false) || a=utf-8

$ echo $? && echo $a
0
utf-8

$ a=$(commandnotfound) || a=utf-8
commandnotfound: command not found

$ echo $? && echo $a
0
utf-8
```

Comment (SuperSandro2000) [reply pkgs/tools/misc/lesspipe/default.nix]: For non Darwin it is probably part of glibc.

Yes I have it enabled but as seen before, it only fails on Darwin without it.

**Review by SuperSandro2000 — APPROVED** [review]
LGTM
